File: src/main/java/grpc/services/LexingCoOrdering/OrderingServer.java
package grpc.services.LexingCoOrdering;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;

import javax.jmdns.JmDNS;
import javax.jmdns.ServiceInfo;
import java.io.IOException;
import java.net.InetAddress;

public class OrderingServer extends LexingCoOrderingServiceGrpc.LexingCoOrderingServiceImplBase {
    private static JmDNS jmDNS;
    private final static int  port = 50052;

    public static void main(String[] args){

        OrderingServer orderingServer = new OrderingServer();
        orderingServer.registerService();
        Server server;
        try {
            server = ServerBuilder.forPort(port).addService(orderingServer).build().start();
            System.out.println("Ordering Server started, listening on " + port);
            server.awaitTermination();
        } catch (IOException | InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void registerService() {

        String service_type = "_ordering._tcp.local.";
        String service_name = "LexingCoOrdering";
        String service_description = "Ordering Server will order parts";

        try {
            //Create JmDNS instance
            jmDNS = JmDNS.create(InetAddress.getLocalHost());

            // Register factory service
            ServiceInfo serviceInfo = ServiceInfo.create(service_type, service_name, port, service_description);
            jmDNS.registerService(serviceInfo);

            System.out.printf("Registered %s in port %d \n", service_type, port);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void sendMessage(MessageRequest request, StreamObserver<MessageReply> responseObserver) {
        System.out.println("Receiving Order request");
        responseObserver.onNext(MessageReply.newBuilder().setText("200 - OK from ordering service").build());
        responseObserver.onCompleted();
    }
}

File: src/main/java/grpc/services/LexingCoOrdering/OrderingServer.kt
package grpc.services.LexingCoOrdering

import grpc.services.LexingCoOrdering.LexingCoOrderingServiceGrpc.LexingCoOrderingServiceImplBase
import io.grpc.Context
import io.grpc.Server
import javax.jmdns.JmDNS
import java.net.InetAddress
import javax.jmdns.ServiceInfo
import java.io.IOException
import io.grpc.stub.StreamObserver
import kotlin.jvm.JvmStatic
import io.grpc.ServerBuilder
import java.lang.InterruptedException

class OrderingServer : LexingCoOrderingServiceImplBase() {

    companion object {
        private const val port = 50052
        @JvmStatic
        fun main(args: Array<String>) {
            val orderingServer = OrderingServer()
            orderingServer.registerService()
            val server: Server
            try {
                server = ServerBuilder.forPort(port).addService(orderingServer).build().start()
                println("Ordering Server started, listening on $port")
                server.awaitTermination()
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }

    private fun registerService() {
        val serviceType = "_ordering._tcp.local."
        val serviceName = "LexingCoOrdering"
        val serviceDescription = "Ordering Server will order parts"
        try {
            val serviceInfo = ServiceInfo.create(serviceType, serviceName, port, serviceDescription)
            //Create JmDNS instance
            val jmDNS: JmDNS? = JmDNS.create(InetAddress.getLocalHost())
            // Register factory service
            jmDNS?.registerService(serviceInfo)

            System.out.printf("Registered %s in port %d \n", serviceType, port)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    //orderStockServerStream takes in the single message from the Warehouse client and
    //returns a message containing the part and quantity that have been ordered
    override fun orderStockServerStream(request: WarehouseRestockRequest, responseObserver: StreamObserver<OrderRestockReply>) {
        println("Receiving warehouse request for new stock")
        val quantity = "5";
        var forked = Context.current().fork()
        var old = forked.attach()
        var preFilteredRequestList = request.text.split("\n").toTypedArray()
        var partsList = preFilteredRequestList.copyOf(preFilteredRequestList.size - 1)

        try {
            for (part in partsList) {
                println("Ordering $quantity new stock of $part")
                responseObserver.onNext(OrderRestockReply.newBuilder().setText("$part\n$quantity").build())
            }
        } finally{
            forked.detach(old)
        }
        responseObserver.onCompleted()
    }
}